source/adapters/level_zero/v2/memory.cpp

```diff
@@ -51,55 +51,81 @@ void *ur_host_mem_handle_t::getPtr(ur_device_handle_t hDevice) {
   return ptr;
 }
 
+ur_result_t ur_device_mem_handle_t::migrateBufferTo(ur_device_handle_t hDevice,
```

Comment:
I don't think we should be doing the same thing as the current adapter. The memory copy should be added to the command list as a dependency for the related enqueue operation.

This whole synchronous migration when setting the parameter seems like a temporary hack that became permanent.

Reply:
It might be the case that just leaving the memory on the device, without migrating between devices is a better option - we should run some benchmarks first. If we decide that we want to migrate the memory, then yes, we should make this asynchronous (and also modify getPtr() function which is used by enqueueKernel, etc.). For now, this function in only used for initialization for copying the data from host to the device - perhaps I should rename the function to avoid confusion?

Reply:
Hm, ok, I see. No, the name is fine, I was just fixated on that scenario.

I think the best approach here would be to create the memory handle in a 'pending' state, to be populated with data asynchronously on first use. My only concern would be concurrency (i.e., what happens when two threads operate on such 'pending' operation at the same time. we'd have to come up with some way to synchronize...).
For now leave it as-is, maybe create a task so that we don't forget.

Reply:
For initialization, we need to do it synchronously I think because the host pointer might not be valid after urMemBufferCreate. We could use some buffers in DRAM (as I did initially) but again, we would probably need to benchmark and see if it makes sense.

For any migration that happens after that, if we just pass all the required dependencies to enqueueMemcpy and either return a signal event or (for in-inorder) just execute that enqueueMemcpy on the commandList that's used for the next operation (kernelLaunch, etc.) we don't need any extra synchronization I think.

Reply:
In general, I guess we could make the logic to decide whether to migrate or not a bit more intelligent that just yes or no 0 - perhaps we could keep some statistics to see how often the memory is accessed from which device and then migrate  (or not) based on that.

Reply:
right, makes sense. let's just do the simple thing and do the copy synchronously. trying to do a copy in dram could make sense for small buffers, but will break down for large things.

The device to device migration path I think may be a bit trickier, because it may involve multiple command lists. We need to start with some tests I think.

Reply:
Yes, we definitely need more tests - I will create a task for that.